sdk/core/Azure.Core/src/Shared/BearerTokenChallengeAuthenticationPolicy.cs

```diff
@@ -0,0 +1,427 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core.Diagnostics;
+
+#nullable enable
+
+namespace Azure.Core.Pipeline
+{
+    /// <summary>
+    /// A policy that sends an <see cref="AccessToken"/> provided by a <see cref="TokenCredential"/> as an Authentication header.
+    /// </summary>
+    internal class BearerTokenChallengeAuthenticationPolicy : HttpPipelinePolicy
+    {
+        private const string AuthenticationChallengePattern = @"(\w+) ((?:\w+="".*?""(?:, )?)+)(?:, )?";
+        private const string ChallengeParameterPattern = @"(?:(\w+)=""([^""]*)"")+";
+        private const string ChallengeHeader = "WWW-Authenticate";
+        private static readonly Regex s_AuthenticationChallengeRegex = new Regex(AuthenticationChallengePattern, RegexOptions.Compiled);
```

Comment:
Regex - even compiled - is pretty expensive. Why not just simple string tokenization? Using spans, it's really cheap. My worry is that this is a hot path depending on the service principal type.

Unfortunately, `StringTokenizer` is not in netstandard2.0, but I effectively ported it here: https://github.com/heaths/stringtokenizer. It's probably worth considering, if not in this PR eventually.

Reply:
I updated my tokenizer benchmarks to do a simple `Regex.Split` and the results are...bad. See https://github.com/heaths/StringTokenizer/blob/83b1a069c017497a185503481dbfdbd3f6970954/README.md#L21. It's 5-10x slower than just `String.Split`, which itself is slow compared to tokenizing strings based on `Span<char>`. It also allocates a lot more memory per call (the regex was compiled). The span method allocates 0 bytes on the heap.

/cc @schaabs

Reply:
Replaced it with a `Span` based implementation. It it 11x faster and allocates significantly less. In addition, it also handles multiple challenges and the quote-less storage challenge format, which the previous implementation didn't. 

```
| Method |       Mean |    Error |   StdDev | Ratio | RatioSD |  Gen 0 |  Gen 1 | Gen 2 | Allocated |
|------- |-----------:|---------:|---------:|------:|--------:|-------:|-------:|------:|----------:|
|  RegEx | 3,520.6 ns | 40.91 ns | 36.27 ns | 11.24 |    0.19 | 0.6638 | 0.0076 |     - |    5553 B |
|   Span |   313.7 ns |  4.63 ns |  4.33 ns |  1.00 |    0.00 | 0.0048 |      - |     - |      40 B |
```